[AOTI] Extend torchgen to generate C shim with version number

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #154848
* __->__ #147745

Summary: While it is ok to add a new arg with defaul value to a fallback op in Python, it will be BC-breaking for the C shim. This PR adds an automatic approach to update C shim files when specifying a version number with a list of new args for the modified op. See https://github.com/pytorch/pytorch/pull/154848 as an example on how to do that.

cc @chenyang78 @penguinwu @yushangdi @benjaminglass1

desertfire committed:
Update

[ghstack-poisoned]

desertfire committed:
Update

[ghstack-poisoned]

desertfire committed:
Update

[ghstack-poisoned]

desertfire committed:
Update

[ghstack-poisoned]

desertfire commented:
@zou3519 , any recommendation on how to fix the BC lint error here?

desertfire commented:
Linking https://dev-discuss.pytorch.org/t/re-generated-aotinductor-c-shim-header-files-have-unexpectedly-changed/2779 for context.

yushangdi commented:
> @zou3519 , any recommendation on how to fix the BC lint error here?

@desertfire If you're sure the bc-linter error changes are safe and expected,  I think you can just add `suppress-api-compatibility-check` or `suppress-bc-linter` to the PR: https://github.com/pytorch/test-infra/wiki/BC-Linter 

```
Error: `Function gen_declaration_and_definition: version_info was added and is now required`
Function gen_c_shim: version_info was added and is now required
Function gen_aoti_c_shim: inductor_fallback_ops was added and is now required
```

zou3519 commented:
@janeyx99 is this related to the version number you wanted?

janeyx99 commented:
@desertfire what ends up consuming the implementations based on version number? Is the version number meant to be user facing? Or internally used by shim.h and used here to differentiate multiple implementations of an op? 

Context: I am planning to provide a global libtorch ABI target version macro, and I'd like to make sure that it should be the same/aligned.

desertfire commented:
> @desertfire what ends up consuming the implementations based on version number? Is the version number meant to be user facing? Or internally used by shim.h and used here to differentiate multiple implementations of an op?
> 
> Context: I am planning to provide a global libtorch ABI target version macro, and I'd like to make sure that it should be the same/aligned.

It's versioning at each op level. See https://dev-discuss.pytorch.org/t/re-generated-aotinductor-c-shim-header-files-have-unexpectedly-changed/2779/4  for an example of how to update some ops with new arguments with default values. If we want to define a global version, we need to think how to compose the story.

desertfire committed:
Update on "[AOTI] Extend torchgen to generate C shim with version number"


Summary: While it is ok to add a new arg with defaul value to a fallback op in Python, it will be BC-breaking for the C shim. This PR adds an automatic approach to update C shim files when specifying a version number with a list of new args for the modified op. TO-BE-FILLED: there will be an example PR linked here later.

cc chenyang78 penguinwu yushangdi benjaminglass1

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Extend torchgen to generate C shim with version number"


Summary: While it is ok to add a new arg with defaul value to a fallback op in Python, it will be BC-breaking for the C shim. This PR adds an automatic approach to update C shim files when specifying a version number with a list of new args for the modified op. See https://github.com/pytorch/pytorch/pull/154848 as an example on how to do that.

[ghstack-poisoned]

desertfire commented:
@pytorchbot merge -f "rocm failure is irrelevant"
